File: app/src/main/java/com/mahmoudmabrok/catechwords/features/displayWords/WordAdapter.java
package com.mahmoudmabrok.catechwords.features.displayWords;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.mahmoudmabrok.catechwords.R;
import com.mahmoudmabrok.catechwords.model.Word;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;


public class WordAdapter extends RecyclerView.Adapter<WordAdapter.Holder> {

    private static final String TAG = "TafseerAdapter";
    private List<Word> list;
    private List<Integer> selected = new ArrayList<>();

    public WordAdapter() {
        list = new ArrayList<>();
    }

    public void clear() {
        list.clear();
        notifyDataSetChanged();
    }

    public List<Word> getList() {
        return list;
    }

    public void setList(List<Word> newList) {
        list = new ArrayList<>(newList);
        notifyDataSetChanged();
    }

    @NotNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.word_item, viewGroup, false);
        Holder holder = new Holder(view);
        return holder;
    }


    @Override
    public void onBindViewHolder(@NonNull Holder holder, int i) {
        Word item = list.get(i);
        holder.tvWord.setText(item.getWord());
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    class Holder extends RecyclerView.ViewHolder {

        @BindView(R.id.tvWord)
        TextView tvWord;

        Holder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

}

File: app/src/main/java/com/mahmoudmabrok/catechwords/features/displayWords/WordAdapter.kt
package com.mahmoudmabrok.catechwords.features.displayWords

import android.os.CountDownTimer
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.mahmoudmabrok.catechwords.R
import com.mahmoudmabrok.catechwords.model.Word
import com.mahmoudmabrok.catechwords.model.isSame
import com.mahmoudmabrok.catechwords.util.log
import java.util.*


class WordAdapter(val listener:IScoreListener) : RecyclerView.Adapter<WordAdapter.Holder>() {
    private var list: ArrayList<Word> = ArrayList()
    private var finished: ArrayList<Int> = ArrayList()
    private var firstSelected = -1
    private var countOfSelected = 0

    private val MAX_SELECTEED = 2
    private val INTERVAL = 300L

    fun setList(newList: ArrayList<Word>) {
        list = ArrayList(newList)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): Holder {
        val view =
            LayoutInflater.from(viewGroup.context).inflate(R.layout.word_item, viewGroup, false)
        return Holder(view)
    }


    override fun onBindViewHolder(holder: Holder, i: Int) {
        "onBindViewHolder $i, first:$firstSelected ".log()
        if (finished.contains(i)) {
            holder.itemView.visibility = View.INVISIBLE
        } else {
            holder.itemView.visibility = View.VISIBLE
            if (i == firstSelected) {
                holder.tvWord.text = list[i].word
            } else {
                holder.tvWord.text = ""
            }
        }

        holder.itemView.setOnClickListener {
            when {
                firstSelected == i -> {
                    alreadyCase(i)
                }

                firstSelected == -1 -> {
                    firstSelected(i)
                }

                firstSelected >= 0 -> {
                    countOfSelected++
                    holder.tvWord.text = list[i].word // place to ui
                    object : CountDownTimer(INTERVAL, INTERVAL) {
                        override fun onFinish() {
                            checkCase(i)
                        }

                        override fun onTick(p0: Long) {
                        }

                    }.start()

                }


            }
        }

        "finished.size == list.size"
/*
        if (finished.size == list.size){
            "finised ${finished.size}"
             listener.onFinish()
        }*/
    }


    /**
     * when click on item that already clicked so ->  hide test
     */
    private fun alreadyCase(i: Int) {
        countOfSelected--
        firstSelected = -1
        notifyItemChanged(i)

    }

    /**
     * first card to be clicked so show it's content
     */
    private fun firstSelected(i: Int) {
        countOfSelected++
        firstSelected = i  // add to firstSelected list
        notifyItemChanged(i)
    }

    /**
     * select second item so we must check
     */
    private fun checkCase(secondIdx: Int) {
        val prev = list[firstSelected]
        val current = list[secondIdx]
        // extension function to test is two cards are correct
        if (current.isSame(prev)) {
            // add ot finished to be hidden next bind
            finished.add(firstSelected)
            finished.add(secondIdx)
            listener.onCorrect()
        } else {
            //todo make vibrate
            "Wrong".log()
        }
        // so when check again  on onBind it will be cleared
        // for second one it is not first index so not displayed at onBind
        firstSelected = -1
        notifyDataSetChanged() //todo find less way
    }


    override fun getItemCount(): Int {
        return list.size
    }

    inner class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var tvWord: TextView = itemView.findViewById(R.id.tvWord)

    }

    interface IScoreListener{
        fun onCorrect()
        fun onFinish()
    }

}